components/camera/stream_subscription.go

```diff
@@ -0,0 +1,126 @@
+package camera
+
+// heavily inspired by https://github.com/bluenviron/mediamtx/blob/main/internal/asyncwriter/async_writer.go
+
+// NOTE: (Nick S)
+// StreamSubscription is what powers camera.SubscribeRTP.
+// It runs a single goroutine which pulls and runs callback functions
+// in order published from a fixed capacity ringbuffer.
+// When the ringbuffer's capacity is reached, `Publish` returns an error.
+// This is important to maintain a bounded
+// amount of queuing as stale video stream packets degrade video quality.
+
+// At time of writing
+// One callback function is created & executed per RTP packet per subscribed client with the nuance that:
+// (at time of writing) gostream is handling multiplexing from 1 track (a camera video feed) to N web
+// browser peer connections interested in that track: https://github.com/viamrobotics/rdk/blob/main/gostream/webrtc_track.go#L126
+// As a result, at time of writing StreamSubscriptions only ever have a single publisher and a single subscriber.
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/bluenviron/gortsplib/v4/pkg/ringbuffer"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"go.viam.com/utils"
+)
+
+// StreamSubscriptionID is the id of a StreamSubscription.
+type StreamSubscriptionID = uuid.UUID
+
+var (
+	// ErrQueueFull indicates the StreamSubscription's queue is full and that
+	// the callback passed to Publish will not be executed.
+	ErrQueueFull = errors.New("StreamSubscription Publish queue full")
+	// ErrClosed indicates the StreamSubscription's is not running.
+	ErrClosed = errors.New("StreamSubscription has been closed")
+	// ErrNegativeQueueSize indicates that the StreamSubscription queueSize
+	// can't be less than 0.
+	ErrNegativeQueueSize = errors.New("StreamSubscription queue size can't be negative")
+)
+
+// StreamSubscription executes the callbacks sent to Publish
+// in a single goroutine & drops Publish callbacks if the
+// buffer is full.
+// This is desirable behavior for streaming protocols where
+// dropping stale packets is desirable to minimize latency.
+type StreamSubscription struct {
+	buffer  *ringbuffer.RingBuffer
```

Comment:
lol, I think that this can be something I take responsibility for figuring out. It is not obvious to me that it is actually a problem that we are using a not-a-ringbuffer here. I'd like to measure the problem before I callout something that might not be a problem